src/main.rs

```diff
@@ -3,8 +3,12 @@ use std::fs;
 use std::io::{self, Write};
 use std::process::ExitCode;
 
+use crate::parser::Parser;
 use crate::scanner::Scanner;
```

Comment:
Btw `Scanner` has a field `source: String`. That's not ideal. In Rust the `String` type is a heap allocated buffer. Every time you copy `String` it will be re-allocated, so you'll store the whole thing again and again and again.

Rust has the `str` type, which acts as a pointer to the originally heap allocated string. Here's where lifetimes get a little tricky because you'll probably end up having to tell Rust that:

```rust
struct Scanner<'the_lifetime_of_the_string> {
  source: &'the_lifetime_of_the_string str, 
}
```

This way you're letting Rust know that `Scanner` can only be around in memory while the String is also around in memory.

Reply:
Right, I figured this was the case but I just wanted this to compile and couldn't be bothered with the lifetimes at the time. I still feel like I don't have the mental modal for lifetimes so I was worried I would create some error that I wouldn't be able to solve.

Reply:
Think about lexical scoping. In JS:

```js
function foo() {
  let bar = {a: 1};
  {
    let baz = {};
    // Do stuff...
  }
  console.log(bar.a)
  // Can't use baz here as it's undefined
  typeof bar == 'undefined'
}

// Can't use bar here, as it's scoped to foo()
typeof bar == 'undefined'
```

Let's look at a psuedo language which had no way of declaring variables and no way handling lexcial scope, instead you had to call a magic `__drop` every time you wanted to free up memory:

```js
function foo() {
  bar = {a: 1};
  baz = {};
  // Do stuff...
  __drop(baz);
  console.log(bar.a)
  __drop(bar);
}
```

This is effectively what lexical scopes do for you. They drop a variable if it leaves a scope. A lexical scope is a _lifetime_. We know `bar` only lives inside of `foo()`. When we finish executing `foo()`, we can delete `bar`. 

However there's a problem. What happens if we give the value to another function:

```js
function foo() {
  let bar = {a: 1);
  someOtherFunction(bar);
}
```

We can no longer delete `bar` as we exit `foo`, because we don't know what `someOtherFunction` is going to do to it. It might send it myriad other places, do async stuff or log it to the console or whatever. In JS we have the garbage collector that keeps track of all these, and knows when to delete it. Rust doesn't have a garbage collector, and so it's difficult to know when it should be deleted. In languages like C/C++ you're often left to your own devices on when this stuff gets freed.

Let's use our pseudo language to introduce a bug to illustrate a failure mode in a language without GC:

```js
function foo() {
  let bar = {a: 1};  // Step 1
  someOtherFunction(bar); // Step 2
  __drop(bar); // Step 4
}
function someOtherFunction(bar) {
  await oneTick(); // Step 3
  bar.a // Step 5 OOPS! `bar` was dropped and so this is now a segfault
}
```

One way this gets solved in languages without GC is to use `Ref` or `RefPtr` classes which count every time they're moved around. Let's introduce something like that to our pseudo language:

```js
class Ref {
  #owners = 1;
  constructor(value) {
    this.value = value;
  }

  release() {
    this.#owners -= 1;
    if (this.#owners < 1) {
      // Here's our magic drop function again
      drop(this.value);
    }
  }

  give() {
    this.#owners +=1;
  }
}

function foo() {
  let bar = new Ref({a:1}); // Step 1; `Ref.#owners == 1`
  someOtherFunction(bar.give()); // Step 2; `Ref.#owners == 2`
  bar.release(); // Step 4; `Ref.#owners ==1`
}

function someOtherFunction(bar) {
  await oneTick(); // Step 3; `Ref.#owners == 2`
  bar.value.a; // Step 5, bar is still in memory as it has 1 owner!
  bar.release(); // Step 6, `Ref.#owners == 0`, `bar` is dropped.
}
```

Here we can see by keeping a bit of memory around (the counter) and by adding some more code, we can be more safe about how many times our value moves around our code, and only free it when nothing else is referencing it.

However, what if we could alter our language/compiler to do all of this for us? What if we could "annotate" variables to alter their lifetimes, letting the compiler reason about these things for us... Let's change our pseudo language and enforce that every variable, parameter, and function has to declare a "lifetime" and our `__drop` function can drop a whole set of variables that belong to a lifetime:

```
function<'A> foo() { // `foo` has an `A` scope/lifetime
 <'A>bar = {a:1};  // `bar` has to live as long as `A`.
  someOtherFunction(bar);
  __drop<'A>(); // Drop all `'A` lifetimes
}

function<'B> someOtherFunction(bar<'B>) { // `someOtherFunction` has the `B` scope/lifetime, and the `bar` variable _has_ to live as long as `B`.
  // Opaque function
  __drop<'B>(); // Drop all `'B` lifetimes
}
```

Now the compiler has some new information to think about. It doesn't know what's going on in `someOtherFunction`, but we've told it that `someOtherFunction` has this `B` lifetime which it keeps for its own execution, and we've attached the `B` to `bar`. So the compiler (again, without thinking about the code inside) knows that `someOtherFunction` _has to keep `bar` around long enough for `someOtherFunction` to finish_. And in this way, we now know we can't call `__drop` on `bar` at the end of `foo`, because the lifetime `L` has to live as long as `O` (and vice versa).

So here, the compiler now has some choices:

 - By giving `bar` to `someOtherFunction`, we give it a new lifetime of `'B` and `__drop<'A>()` won't drop `bar`. Instead, `bar` now lives as long as `'B` and gets dropped when `__drop<'B>();` is called.
 - Throw a compile error telling the user that they need to do other special things to `bar` so it can be safely passed to `someOtherFunction`.

Rust takes this a step further, because of borrowing and mutability. However those topics - while adjacent - are different enough that they don't _need_ to be discussed as part of lifetimes.

Reply:
> A lexical scope is a _lifetime_.

> Rust takes this a step further, because of borrowing and mutability. However those topics - while adjacent - are different enough that they don't _need_ to be discussed as part of lifetimes.

These two paragraphs a long with the very excellent code examples make me feel like I understand lifetimes now.

I can probably take that knowledge and refactor the code now to use lifetimes where applicable.

Thanks ✨

Reply:
Rust has two "special" reserved lifetimes: `'static` which means "the life of the executable". This tells the compiler this variable will always be present in memory. The other is the "placeholder" lifetime `'_` which tells the compiler to "just figure it out".

All other lifetimes are made up. `'a` is kind of like `<T>`. It has become popular lifetime name but it holds no meaning.

Reply:
The scanner is now using `&'a str` for the source code. I might have added more errors but hopefully I'm going in the right direction.

Reply:
Yep `&'a str` sounds perfect.